geometric_kernels/utils/utils.py

```diff
@@ -289,3 +300,169 @@ def get_resource_file_path(filename: str):
     else:
         with impresources.path(resources, filename) as path:
             yield path
+
+
+def hamming_distance(x1: B.Bool, x2: B.Bool):
+    """
+    Hamming distance between two batches of boolean vectors.
+
+    :param x1:
+        Array of any backend, of shape [N, D].
+    :param x2:
+        Array of any backend, of shape [M, D].
+
+    :return:
+        An array of shape [N, M] whose entry n, m contains the Hamming distance
+        between x1[n, :] and x2[m, :]. It is of the same backend as x1 and x2.
+    """
+    return count_nonzero(logical_xor(x1[:, None, :], x2[None, :, :]), axis=-1)
+
+
+def log_binomial(n: B.Int, k: B.Int) -> B.Float:
+    """
+    Compute the logarithm of the binomial coefficient.
+
+    :param n:
+        The number of elements in the set.
+    :param k:
+        The number of elements to choose.
+
+    :return:
+        The logarithm of the binomial coefficient binom(n, k).
+    """
+    assert B.all(0 <= k <= n)
+
+    return B.loggamma(n + 1) - B.loggamma(k + 1) - B.loggamma(n - k + 1)
+
+
+def binary_vectors_and_subsets(d: int):
+    r"""
+    Generates all possible binary vectors of size d and all possible subsets of
+    the set $\{0, .., d-1\}$ as a byproduct.
+
+    :param d:
+        The dimension of binary vectors and the size of the set to take subsets of.
+
+    :return:
+        A tuple (x, combs), where x is a matrix of size (2**d, d) whose rows
+        are all possible binary vectors of size d, and combs is a list of all
+        possible subsets of the set $\{0, .., d-1\}$, each subset being
+        represented by a list of integers itself.
+    """
+    x = np.zeros((2**d, d), dtype=bool)
+    combs = []
+    i = 0
+    for level in range(d + 1):
+        for cur_combination in combinations(range(d), level):
+            indices = list(cur_combination)
+            combs.append(indices)
+            x[i, indices] = 1
+            i += 1
+
+    return x, combs
+
+
+def np_to_backend(value: B.NPNumeric, backend: str):
+    """
+    Converts a numpy array to the desired backend.
+
+    :param value:
+        A numpy array.
+    :param backend:
+        The backend to use, one of the strings "tensorflow", "torch", "numpy", "jax".
+
+    :raises ValueError:
+        If the backend is not recognized.
+
+    :return:
+        The array `value` converted to the desired backend.
+    """
+    if backend == "tensorflow":
+        import tensorflow as tf
+
+        return tf.convert_to_tensor(value)
+    elif backend in ["torch", "pytorch"]:
+        import torch
+
+        return torch.tensor(value)
+    elif backend == "numpy":
+        return value
+    elif backend == "jax":
+        import jax.numpy as jnp
+
+        return jnp.array(value)
+    else:
+        raise ValueError("Unknown backend: {}".format(backend))
+
+
+def array_type(backend: str):
```

Comment:
This and the function below look quite hacky. We have something similar going on in `tests/test_dtype.py` which is justified. I see less justification for this in the "main" code. In any case, this seems to be only required in some tests. Consider moving it there and/or reusing the existing code where possible.

Reply:
I moved these into `tests/helper.py`.